src/Neo/Ledger/Blockchain.cs

```diff
@@ -152,17 +154,32 @@ public Blockchain(NeoSystem system)
 
         private void OnImport(IEnumerable<Block> blocks, bool verify)
         {
+            var blocksToImport = blocks as Block[] ?? blocks.ToArray();
+            var hash = blocksToImport.FirstOrDefault()?.NextConsensus;
+
+            // Capture necessary context elements
+            var self = Self;
+            var verificationTasks = blocksToImport.Select(block => Task.Run(() =>
+            {
+                if (verify && !block.Header.VerifyWitness(system.Settings, system.StoreView, hash, ((IVerifiable)block.Header).Witnesses[0], 3_00000000L, out long fee))
```

Comment:
you only verify the witness, we need to call `block.verify`

Reply:
Block verify is for new blocks, I checked the logic, the core is this header witness verify.

Reply:
Can you add a fallback? Let's say we find a problem that wasn't expected in production. Can you add in single threading options _(the current way it is now)_.

Reply:
I dont think a fallback will work, currently we dont have exception handling for erronious importing. Maybe in another pr.

Reply:
> Block verify is for new blocks, I checked the logic, the core is this header witness verify.

We need to ensure that the previous hash is the same, etc...

Reply:
Maybe we can have two verify import modes, full and only signatures, but we require the full version @Jim8y

Reply:
Those are redundant check. You have valid witness check of a block with the correct index. The only chance to have an incorrect prehash is the dbft consensus is broken.

Reply:
Not true, `witness` only verifies `witness` section, which is address `ScriptHash` and `InvocationScript` and `VerificationScript`. thats it.

Reply:
> Those are redundant check. You have valid witness check of a block with the correct index. The only chance to have an incorrect prehash is the dbft consensus is broken.

no, you can download a wrong file, if you want to verify it you should be able to do that

Reply:
As i said, its the same:

```C#
        internal bool Verify(ProtocolSettings settings, DataCache snapshot)
        {
            if (primaryIndex >= settings.ValidatorsCount)
                return false;
            TrimmedBlock prev = NativeContract.Ledger.GetTrimmedBlock(snapshot, prevHash);
            if (prev is null) return false;
            if (prev.Index + 1 != index) return false;
            if (prev.Header.timestamp >= timestamp) return false;
            if (!this.VerifyWitnesses(settings, snapshot, 3_00000000L)) return false;
            return true;
        }
```

Reply:
If you still think a verification is necessary, i can add another pr to fix it.

Reply:
If it's the same, why not call it?

Reply:
Be professional please, as i have explained, the checking logic is redundant and unnecessary, primaryindex check is meanless cause you have valid consensus signatures, prevblock is not necessary cause you already have it during the importing process, timestamp the same, a valid witness ensures everything in the block valid as long as its at the right order. 

> We need to ensure that the previous hash is the same, etc...

This is your original question, as i explined, the original code has no verification at all, so i will not change that logic but in another pr.

Reply:
> If it's the same, why not call it?

But this code does contain bugs, I am still working to fix them. And i think i may still make it much much faster.